Improve localization (#2111)
Adyen/adyen-ios: Adyen.docc/Localization.md

# Localization

Both the Drop-in and the Components offer an option to customize the strings to match your app's use case or tone of voice.

By default, the SDK attempts to use a device's locale for translation of text and formatting of monetary values. If the preferred device locales are not supported, the SDK falls back to the **en-US** locale.

> Note: Localization only picks up locales that are listed in the `CFBundleLocalizations` property of your app's `Info.plist` file.

> ## Enforcing locale

Reviewer: Enforce **a** locale

> ## Enforcing locale

Reviewer: ```suggestion
## Enforce a locale
```

> To enforce a specific locale and formatting of monetary values, regardless of the shopper's device locale, use `LocalizationParameters(enforcedLocale: MY_LOCALE)`. (`MY_LOCALE` should be replaced with your desired locale identifier, e.g., "fr-FR").

Reviewer: ```suggestion
To enforce a specific locale and formatting of monetary values, regardless of the shopper's device locale, use `LocalizationParameters(enforcedLocale: MY_LOCALE)`: 
-  Replace `MY_LOCALE` with your desired locale identifier, for example, **fr-FR**.
```

> ## Overriding default formatting of monetary values

Reviewer: ```suggestion
## Override default monetary value formats
```

> To enforce a custom locale for formatting a monetary values, use the `locale` property on `LocalizationParameters`. In case of `enforcedLocale` the value will be used for both localisation and monetary values formatting.

Reviewer: ```suggestion
To enforce a custom locale for the formatting of monetary values, use the `locale` property on `LocalizationParameters`. If you are enforcing a locale, the locale you are enforcing is used for both localization and the formatting of monetary values.
```

> ## Overriding strings

Reviewer: ```suggestion
## Override strings
```

You can override strings for each key and locale.

> > Important: If you are using multiple Adyen Components or the Drop-in, all usages of that localization key will be updated with the overridden value.

Reviewer: overridden value would be the one that is being overridden by what they specify, to prevent confusion, it could be better to say 
"all usages of that localization key will be updated with the value you specify to override the default localization."

> > Important: If you are using multiple Adyen Components or the Drop-in, all usages of that localization key will be updated with the overridden value.

Reviewer: ```suggestion
> Important: If you are using multiple Adyen Components or the Drop-in, all usages of that localization key will be updated with the value you specify to override the default localization.
```

> Tip: If you want to override the value of a key for a specific component, you can use the `localizationParameters` property to specify a custom localization file to get the values from, as described below.

1.  [Add a string catalog to your project](https://developer.apple.com/documentation/xcode/localizing-and-varying-text-with-a-string-catalog#Add-a-string-catalog-to-your-project) or use your existing one.
2.  [Add a language to your project](https://developer.apple.com/documentation/xcode/localizing-and-varying-text-with-a-string-catalog#Add-a-language-to-your-project) if necessary.
3.  Find the key for the string you want to translate in the [list of available strings](https://github.com/Adyen/adyen-ios/blob/develop/Adyen/Assets/Generated/LocalizationKey.swift) and override it for each desired locale.

> For example, if you want to override the payment button text to **Subscribe for [AMOUNT]**, add a key `adyen.submitButton.formatted` to your catalog. Then, provide new translation(s).

Reviewer: ```suggestion
For example, if you want to override the payment button text to **Subscribe for [AMOUNT]**: 
1. Add the `adyen.submitButton.formatted` key to your catalog.
2. Provide the new translation(s).
```

# Localization

> The SDK offers two primary ways to handle language and regional formatting:

Author: ```suggestion
The SDK allows you to customize UI strings and manage locale behavior using the `LocalizationParameters` object. You can configure these settings for both the Drop-in and individual Components through their `configuration.localizationParameters` property.

Based on this configuration, the SDK provides two methods for handling language and regional formatting:
```

* **[iOS Default](https://developer.apple.com/library/archive/qa/qa1828/_index.html) Localization:**

>     The SDK tries to match the shopper's device language and regional settings. This is effective only for locales also listed in your app's `CFBundleLocalizations` array (found in the `Info.plist` file). If a suitable match for UI text isn't found, the text defaults to `en-US`, though formatting for amounts or dates may still try to adhere to the shopper's device region.

Reviewer: ```suggestion
    The SDK matches the shopper's device language and regional settings when their locale is listed in your app's `CFBundleLocalizations` array in the `Info.plist` file. If the locale for your shopper's device language isn't found, the locale defaults to `en-US` for UI text. Amount and date formatting may still adhere to the shopper's device region.
```

* **Enforced Localization:**

>     You can explicitly set a specific locale (e.g., `"fr-FR"`) by using `LocalizationParameters(enforcedLocale: "fr-FR"`. This ignores device preferences and ensures that all UI text and data formatting (like currency, dates, numbers) consistently use the specified locale's conventions. This method is ideal for apps targeting a single language, requiring uniform branding, or implementing an in-app language switcher.

Reviewer: ```suggestion
    You enforce a specific locale (e.g., `"fr-FR"`) by using `LocalizationParameters(enforcedLocale: "fr-FR"`. This ignores the shopper's device preferences, and all UI text and data formatting (like currency, dates, numbers) use the conventions of the enforced locale. Using an enforced locale allows you to target a single language, achieve uniform branding, or implement an in-app language switcher.
```

> ## How Device Locales Affect Monetary Formatting

Reviewer: ```suggestion
## How device locales affect monetary formatting
```

> The default iOS localisation can lead to significant variations in the display of monetary values based on the shopper's device locale settings.

Reviewer: ```suggestion
The default iOS localization can lead to significant variations in how monetary values are displayed based on the shopper's device locale settings.
```

> For instance, if your app's UI supports English, French, and Chinese (Mainland) (meaning all three are listed in your `CFBundleLocalizations`), and the transaction amount is **12,340 Chinese Yuan (CNY)** (twelve thousand three hundred forty), here's how four different shoppers might see that CNY amount displayed, formatted according to their device's regional settings:

Reviewer: ```suggestion
For instance, if your app's `CFBundleLocalizations` array consist of English, French, and Chinese (Mainland) locales, and the transaction amount is **12,340 Chinese Yuan (CNY)** (twelve thousand three hundred forty), below is how your app will display the amount, formatted according to their device's regional settings:
```

| User's Device Language | User's Device Region | Example Pay Button Label (for 12,340 CNY) |
| :--------------------- | :------------------- | :---------------------------------------- |
| `zh` (Chinese)         | `CN` (China)         | `¥12,340.00`                              |
| `fr` (French)          | `FR` (France)        | `12 340,00 CNY`                           |
| `tr` (Turkish)         | `TR` (Turkey)        | `CN¥12.340,00`                            |
| `vi` (Vietnamese)      | `VN` (Vietnam)       | `¥12.340,00`                              |

> ## Uniform Monetary Formatting

Reviewer: ```suggestion
## Uniform currency and amount formatting
```

General question, does monetary formatting sound good to you? Can we just say currency and amount formatting to use simpler language?

> If you prefer a consistent monetary display for all shoppers, or need to customize how numbers are formatted independently of the UI language, you can use `LocalizationParameters`. There are two main ways to control this:

Reviewer: ```suggestion
If you prefer to have a uniform monetary formatting for all shoppers, or need to customize how numbers are formatted independently of the UI language, you can use `LocalizationParameters`. There are two ways to control this:
```

* **To customize *only* the formatting of monetary values (numbers, separators):**

>     Use the `locale` property on `LocalizationParameters`. This applies the monetary formatting of the chosen locale, but the UI text will still follow the "Natural Localization" process.

Reviewer: ```suggestion
    Use the `locale` property on `LocalizationParameters`. This applies the monetary formatting of the chosen locale, but the UI text will adhere to the "Default Localization".
```

> * **To enforce a single locale for *both* UI text and all data formatting (including monetary values):**

Author: ```suggestion
* **To enforce a single locale for *both* UI text and numeric data formatting:**
```

>     Use the `enforcedLocale` property on `LocalizationParameters` (as detailed in the "Enforced Localization" section above).## Override strings

Reviewer: ```suggestion
    Use the `enforcedLocale` property on `LocalizationParameters` (as detailed in the "Enforced Localization" section above).
```

## Override strings

You can override strings for each key and locale.

> Important: If you are using multiple Adyen Components or the Drop-in, all usages of that localization key will be updated with the value you specify to override the default localization.

> > Tip: If you want to override the value of a key for a specific component, you can use the `localizationParameters` property to specify a custom localization file to get the values from, as described below.

Reviewer: is this `localizationParameters` different than the one above which is `LocalizationParameters`, first one is camel case, just to check for consistency?

Author: I have added an explanation to the first  paragraph.

1.  [Add a string catalog to your project](https://developer.apple.com/documentation/xcode/localizing-and-varying-text-with-a-string-catalog#Add-a-string-catalog-to-your-project) or use your existing one.
2.  [Add a language to your project](https://developer.apple.com/documentation/xcode/localizing-and-varying-text-with-a-string-catalog#Add-a-language-to-your-project) if necessary.

> 3.  Find the key for the string you want to translate in the [list of available strings](https://github.com/Adyen/adyen-ios/blob/develop/Adyen/Assets/Generated/LocalizationKey.swift) and override it for each desired locale.

Reviewer: ```suggestion
3.  Find the key for the string you want to translate in the [list of available strings](https://github.com/Adyen/adyen-ios/blob/develop/Adyen/Assets/Generated/LocalizationKey.swift) and add your custom translation to override it for each desired locale.
```

You can override strings for each key, and for each language and locale using legacy `.strings` files.

1.  In Xcode, [create a new](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/LoadingResources/Strings/Strings.html) or use your existing `Localizable.strings` file. (If you are using `.xcstrings` catalogs, refer to the section above).

    For example, if your app uses English and Spanish, your project folder should have a `Localizable.strings` file for each locale:

    - English: `en-US.lproj/Localizable.strings`
    - Spanish: `es-ES.lproj/Localizable.strings`

2.  Find the key for the string you want to translate in the [list of available strings](https://github.com/Adyen/adyen-ios/blob/develop/Adyen/Assets/Generated/LocalizationKey.swift) and override it for each desired locale.

For example, if you want to override the payment button text to **Subscribe for [AMOUNT]**:

- English, in the `en.lproj/Localizable.strings` file:

  ```strings
  "adyen.submitButton.formatted" = "Subscribe for %@";
  ```

- Spanish, in the `es.lproj/Localizable.strings` file:
  ```strings
  "adyen.submitButton.formatted" = "Suscríbete por %@";
  ```

</details>

> ### Adding a New Locale

Reviewer: ```suggestion
### Add a new locale
```

| Spanish                | es-ES       |          |
| Swedish                | sv-SE       |          |

> ### Custom Localization File Name

Reviewer: ```suggestion
### Use a custom localization file name
```
